src/messaging/ReliableMessageMgr.cpp

```diff
@@ -106,138 +94,62 @@ void ReliableMessageMgr::TicklessDebugDumpRetransTable(const char * log)
 
 void ReliableMessageMgr::ExecuteActions()
 {
+    System::Clock::Timestamp now = System::SystemClock().GetMonotonicTimestamp();
+
 #if defined(RMP_TICKLESS_DEBUG)
-    ChipLogDetail(ExchangeManager, "ReliableMessageMgr::ExecuteActions");
+    ChipLogDetail(ExchangeManager, "ReliableMessageMgr::ExecuteActions at % " PRIu64 "ms", now.count());
 #endif
 
-    ExecuteForAllContext([](ReliableMessageContext * rc) {
+    ExecuteForAllContext([&](ReliableMessageContext * rc) {
         if (rc->IsAckPending())
         {
-            if (0 == rc->mNextAckTimeTick)
+            if (rc->mNextAckTime <= now)
             {
 #if defined(RMP_TICKLESS_DEBUG)
-                ChipLogDetail(ExchangeManager, "ReliableMessageMgr::ExecuteActions sending ACK");
+                ChipLogDetail(ExchangeManager, "ReliableMessageMgr::ExecuteActions sending ACK %p", rc);
 #endif
-                // Send the Ack in a SecureChannel::StandaloneAck message
                 rc->SendStandaloneAckMessage();
             }
         }
     });
 
-    TicklessDebugDumpRetransTable("ReliableMessageMgr::ExecuteActions Dumping mRetransTable entries before processing");
-
-    // Retransmit / cancel anything in the retrans table whose retrans timeout
-    // has expired
+    // Retransmit / cancel anything in the retrans table whose retrans timeout has expired
     mRetransTable.ForEachActiveObject([&](auto * entry) {
-        CHIP_ERROR err = CHIP_NO_ERROR;
-
-        if (entry->nextRetransTimeTick != 0)
+        if (entry->nextRetransTime > now)
             return true;
 
-        if (entry->retainedBuf.IsNull())
-        {
-            // We generally try to prevent entries with a null buffer being in a table, but it could happen
-            // if the message dispatch (which is supposed to fill in the buffer) fails to do so _and_ returns
-            // success (so its caller doesn't clear out the bogus table entry).
-            //
-            // If that were to happen, we would crash in the code below.  Guard against it, just in case.
-            ClearRetransTable(*entry);
-            return true;
-        }
+        VerifyOrDie(!entry->retainedBuf.IsNull());
 
         uint8_t sendCount       = entry->sendCount;
         uint32_t messageCounter = entry->retainedBuf.GetMessageCounter();
 
         if (sendCount == CHIP_CONFIG_RMP_DEFAULT_MAX_RETRANS)
         {
-            err = CHIP_ERROR_MESSAGE_NOT_ACKNOWLEDGED;
-
             ChipLogError(ExchangeManager,
                          "Failed to Send CHIP MessageCounter:" ChipLogFormatMessageCounter " on exchange " ChipLogFormatExchange
                          " sendCount: %" PRIu8 " max retries: %d",
                          messageCounter, ChipLogValueExchange(&entry->ec.Get()), sendCount, CHIP_CONFIG_RMP_DEFAULT_MAX_RETRANS);
 
-            // Remove from Table
-            ClearRetransTable(*entry);
+            // Do not StartTimer, we will schedule the timer at the final end
```

Comment:
At the final end of what?  I don't see it happening later in this function.  Please document where the timer is actually started?

Reply:
Update the comment